design/tls-backend-verification.md

```diff
@@ -0,0 +1,166 @@
+# TLS backend verification
+
+Status: Draft
+
+Contour 0.11 added the ability to communicate via TLS between Envoy and pods in the backend service.
+This proposal describes the facility for Envoy to verify the backend service's certificate.
+
+## Goals
+
+- Allow IngressRoute authors to assert that the backend service Envoy is communicating with is using a certificate signed by a certificate authority of their choice.
+- Allow IngressRoute authors to assert that the backend service Envoy is communicating with presents a subjectAltName of their choice.
+
+## Non Goals
+
+- Authentication via client side certificates is not in scope.
+- Support for k8s ingress documents.
+
+## Background
+
+Contour 0.11 added the `contour.heptio.com/upstream-protocol.tls` Service annotation which told Contour to use TLS when communicating with the members of the backend cluster, the Service's pods.
+However, this connection is not validated, Envoy will accept any certificate the client presents.
+
+This proposal seeks to improve the validation of TLS connections .
+
+## High-Level Design
+
+At a high level I propose the following:
+
+1. The CA to be used to validate the certificate presented by the backend should be packaged in a Secret.
+The secret will be placed in the namespace of the IngressRoute.
```

Comment:
Is there anything a configmap does better than a secret for this use-case?

Reply:
They don't do anything that secrets don't do, but you could use the same logic to say that Kubernetes shouldn't support configmaps at all.

It looks like the reasons for implementing it with secrets are about discouraging abuse of configmaps for other use cases. It's your code base so I don't have any strong opinions on that front, just wanted to make a note that CA certificates generally shouldn't be considered private material.

Reply:
Gotcha, thanks for clarifying @robbiemcmichael. I appreciate the review on this! =)

Reply:
I agree that a CA is not private material. The justification for using a secret is contour is already watching all secrets.

Storing this information in a config map means contour needs to watch all config maps, adding load to the API server. Additionally I'm very scared that this will create a precedent for storing arbitrary contour configuration in one or more config maps. I do not want to do this unless absolutely necessary and only after rigourous investigation of other options.